### ISSUE TITLE
Attribute to disallow usage of positional parameters (avoid breaking change in source generated code)

### ISSUE THREAD

**Issue by joshlang** [dotnet/roslyn#60263]
## Background and Motivation

TLDR:  A **[DisallowPositionalParameters]** method attribute to avoid source generators creating (unnoticed) breaking changes.

Consider a source generator that transforms:

```
partial class Stuff
{
    public string SomeProperty { get; init; }
    public string AnotherProperty { get; init; }
}
```

And takes the properties and creates something like:

```
partial class Stuff
{
    void DoSomething(string someProperty = "", string anotherProperty = "") {...}
}
```

which is consumed like:  `.DoSomething("some property", "another property")`

...

And now, the developer adds a new property to the class

```
partial class Stuff
{
    public string SomeProperty { get; init; }
    public string NewProperty { get; init; } /// <-- the new property
    public string AnotherProperty { get; init; }
}
```

Which changes the generated code to:

```
partial class Stuff
{
    void DoSomething(string someProperty = "", string newProperty = "", string anotherProperty = "") {...}
    // ^-- The parameters are in a different order, but of the same type, and with default values!
}
```

Now, the consuming code is **BROKEN**.  The order of parameters is not stable.

There does not appear to be any convenient way, in source generators, to account for these changes and avoid creating breaking changes.

Even if the original class changed, extra care would have needed to be taken to make sure to output the method parameters in a stable, unchanging order.

However, the consumer is immune from the problem (and source generators don't need to worry any more) if they use *named* parameters instead of positional parameters.

It doesn't mean the consuming code won't be broken - only that it won't be broken without being noticed.

## Proposed API

```
[AttributeUsage(AttributeTargets.Method)]
public sealed class DisallowPositionalParameters : Attribute {}
```

And now the compiler will throw an error if a method uses positional parameters

And ideally, perhaps an analyzer could auto-change positional into named parameters for convenience

At runtime, positional parameters would still be allowed and `[DisallowPositionalParameters]` would have no effect

## Usage Examples

The source generator above would simply output
```
partial class Stuff
{
    [DisallowPositionalParameters]
    void DoSomething(string someProperty = "", string newProperty = "", string anotherProperty = "") {...}
}
```

## Alternative Designs

Perhaps a warning instead of error.

## Risks

I'm sure someone will think of something for this section!

**Comment by CyrusNajmabadi** [comment]
Note: there's no need for roslyn to do anything here.  You can already create your own attribute and analyzer with whatever restriction rules you want.  That's one of the use cases for analyzers :)

**Comment by joshlang** [comment]
Yeah, already doing that.

But I figured it was enough of an issue, especially given the push towards source generation, that maybe there'd be interest in making it "official"

Sounds like... not :)
